add support for 100-continue for H/2

wfurt opened the pull request:
fixes #31312 
This changes adds logic to process Expect: 100-continue request header and 10x responses.

davidsh commented on src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.cs:
```diff
@@ -1878,11 +1878,14 @@ public async Task PostAsync_CallMethod_EmptyContent(Uri remoteServer)
+        [InlineData(false, "2.0")]
+        [InlineData(true, "2.0")]
+        [InlineData(null, "2.0")]
         public async Task PostAsync_ExpectContinue_Success(bool? expectContinue, string version)
```
I don't understand this test.  Why is the parameter for 'expectContinue' a nullable-bool?

wfurt replied:
I can look into it. This is existing test I updated for 2.0

stephentoub replied:
> Why is the parameter for 'expectContinue' a nullable-bool?

To validate the behavior when you explicitly set it to true (expectContinue==true), explicitly set it to false (expectContinue==false), or don't set it to anything explicitly and just have the default (expectContinue==null).

davidsh replied:
I still don't really understand this.  A request either sets the 'Expect: 100-continue' header or it doesn't set the header at all.  Are you saying the test tries to validate cases where the header is 'Expect: somethingelse' ?

wfurt replied:
Seems like ExpectContinue is nullable bool?
So this seems to verify default behavior when is it is not explicitly set.

stephentoub replied:
> So this seems to verify default behavior when is it is not explicitly set.

Exactly.  Did the developer explicitly set the ExpectContinue header to true, did they explicitly set the ExpectContinue header to false, or did they not touch it.

davidsh replied:
I guess I'm confused by this because I thought the developer added the header manually.  It sounds like you think they call a property on the HttpRequestMessage that is a bool?

stephentoub replied:
https://github.com/dotnet/corefx/blob/a10890f4ffe0fadf090c922578ba0e606ebdd16c/src/System.Net.Http/src/System/Net/Http/Headers/HttpRequestHeaders.cs#L69

davidsh replied:
Ok thanks for the reminders of how this property/header works.

In hindsight, this property should probably have been a normal bool.  Where true' means to emit the 'Expect: 100-continue' request header and 'false' means to emit no header.  The RFC doesn't define any other value for the 'Expect' header.

davidsh commented on src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.cs:
```diff
@@ -1878,11 +1878,14 @@ public async Task PostAsync_CallMethod_EmptyContent(Uri remoteServer)
             using (HttpClient client = CreateHttpClient())
             {
-                var req = new HttpRequestMessage(HttpMethod.Post, Configuration.Http.RemoteEchoServer)
+                var req = new HttpRequestMessage(HttpMethod.Post, version == "2.0" ? Configuration.Http.Http2RemoteEchoServer : Configuration.Http.RemoteEchoServer)
```
The remote Azure echo servers don't really do anything with 'Expect: 100-continue'.  They never will send back a '100 Continue' HTTP response status code.

Normally, the whole point of even using 'Expect: 100-continue' semantics is to help avoid POST'ing a large request body to the server in case the server demand authentication.  So, using this feature allows the client to wait before sending a large request body before getting denied with a '401 Unauthorized' for example.

So, I'm not clear about the purpose of these tests here using the remote Azure echo servers.  At most, it is testing for nothing bad happening when sending 'Expect: 100-continue' and not getting any 100 Continue response. 

It might be useful to add comments to this test to explain that. Otherwise, people reading this code might assume that it is testing something that it really isn't.

wfurt replied:
I'm not 100% sure either but #34460 asked for updating all existing tests to external servers to also use HTTP/2. As minimum we will get validation that we can interoperate with real server and not just loopback.

davidsh replied:
ok

Tratcher replied:
100 Continue responses are automatic as soon as you read the request body. I assume the echo server does that?

davidsh replied:
That's a great question.  I'm not sure if the echo server really support 100 Continue at all.

The "echo" server is running as an ASP.NET Core 2.2 server running in Azure App Service:
https://github.com/davidsh/corefx-net-endpoint/tree/master/NetCoreServer.  

I think the default of ASP.NET is IIS?  Does that support 100 Continue by default?

Tratcher replied:
Yes, I think all MS servers support 100 Continue. It's implemented in Http.Sys so HttpListener, IIS, and HttpSys (WebListener) all get it for free. Kestrel also supports it. There is no application code required to send 100 Continue, only reading from the request body.

davidsh commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs:
```diff
@@ -108,6 +109,26 @@ public async Task SendRequestBodyAsync(CancellationToken cancellationToken)
+                            allowExpect100ToContinue, _connection._pool.Settings._expect100ContinueTimeout, Timeout.InfiniteTimeSpan);
+                Task response = ReadResponseHeadersAsync(true, allowExpect100ToContinue);
+
+                // By now, either we got response from server or timer expired.
```
nit: I think this would be better phrased as "Wait until we either get a response from server or timer expires."

davidsh commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs:
```diff
@@ -195,6 +216,21 @@ public void OnResponseHeadersStart()
 
+            public void OnResponse100Continue()
+            {
+                // This is called when we receive complete set of headers with transient response code - like 100 Continue.
```
"with transient response code - like 100 Continue".  This implies that this method could be called with other interim 1xx responses.  I don't think that is really true with this code.  So, I would rephrase this comment.

wfurt replied:
this is more about server sending us unexpected response. When I look at HTTP/1 it consumes all 10x so I did it as well. We could possibly throw protocol error otherwise.

davidsh replied:
So, this method is named "OnResponse100Continue()" but it can be called with any 1xx interim response received from the server?  If so, the method should be renamed to something like "OnInterimResponseReceived()".

wfurt replied:
I checked and we do process all 1xx responses. However, this function is called only for 100 responses.   I updated the comment.

davidsh commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs:
```diff
@@ -323,15 +359,38 @@ public void OnResponseAbort()
+                    if (Response.StatusCode  == HttpStatusCode.Continue)
+                    {
+                        OnResponse100Continue();
+                        if (tcs != null)
```
This could be change to?

```c#
tcs?.TrySetResult(true);
```

davidsh commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs:
```diff
@@ -570,6 +630,19 @@ public override ValueTask WriteAsync(ReadOnlyMemory<byte> buffer, CancellationTo
+
+                    if (http2Stream._abortRequestBody)
+                    {
+                        // If are asked to abort sending request body after we started, send RST and ignore test for the stream.
```
nit:  should be something like "If asked to abort sending request body after we started, send RST and ignore test for the stream."

And should this "ignore test for the stream" be changed to "ignore the rest of the stream"?  I don't know enough about what this code is doing.

wfurt commented:
all tests are finally passing.

davidsh commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs:
```diff
@@ -107,6 +108,27 @@ public async Task SendRequestBodyAsync(CancellationToken cancellationToken)
+
+                // By now, either we got response from server or timer expired.
+                bool sendRequestContent = await allowExpect100ToContinue.Task.ConfigureAwait(false);
+                expect100Timer.Dispose();
```
It might be better to wrap this in a 'using' for the timer instead of having an explicit call to .Dispose(). Doing so would allow it to run in an implied 'finally' block.  So, if the code inside the block throws an exception, the timer will definitely be explicitly Dispose'd.

wfurt replied:
updated.

geoffkizer commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:
```diff
@@ -1344,13 +1345,23 @@ public sealed override async Task<HttpResponseMessage> SendAsync(HttpRequestMess
 
-                // Wait for response headers to be read.
-                await http2Stream.ReadResponseHeadersAsync().ConfigureAwait(false);
+                if (!hasExpectContinueHeader )
```
nit: remove space before closing paren

geoffkizer commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs:
```diff
@@ -322,15 +359,35 @@ public void OnResponseAbort()
+                }
+                while ((uint)Response.StatusCode < 200);
+
+                if (tcs != null && (int)Response.StatusCode >= 300)
```
It seems like the tcs should be set to null when we successfully receive a 100 Continue response. Am I missing where this happens?

wfurt replied:
I don't understand. We may not have received 100 or we may not do it at all. I simplified the section.

geoffkizer commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs:
```diff
@@ -322,15 +359,35 @@ public void OnResponseAbort()
             }
 
-            public async Task ReadResponseHeadersAsync()
+            public async Task ReadResponseHeadersAsync(bool createStream, TaskCompletionSource<bool> tcs = null)
```
`createStream` is unused?

geoffkizer commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs:
```diff
@@ -107,6 +108,27 @@ public async Task SendRequestBodyAsync(CancellationToken cancellationToken)
                 }
             }
 
+            public async Task SendRequestContentWithExpect100ContinueAsync(CancellationToken cancellationToken)
```
Does this method have to live on `Http2Stream`, or can it live on `Http2Connection`? Most of the actual send/response logic is currently on `Http2Connection`, and it might be better to have this method there too. The `Http2Stream` logic is intentionally more low-level.

Similarly, below, we are looping in `ReadResponseHeadersAsync` when we get a 1xx response. I wonder if it would be better to leave `ReadResponseHeadersAsync` as-is and add logic in the caller to deal with 1xx responses, including 100 Continue.

geoffkizer replied:
Related to this, it seems like we don't need a separate flag for `_abortRequestBody` if the caller handles response codes itself. Just check for a >= 300 response code and don't send the body in that case.

wfurt replied:
I put it there because that where ReadResponseHeadersAsync() and SendRequestBodyAsync(). The 100 is specific to each request and mapping to stream seems more appropriate.

geoffkizer commented on src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.cs:
```diff
@@ -1834,11 +1834,14 @@ public async Task PostAsync_CallMethod_EmptyContent(Uri remoteServer)
+        [InlineData(false, "2.0")]
+        [InlineData(true, "2.0")]
+        [InlineData(null, "2.0")]
         public async Task PostAsync_ExpectContinue_Success(bool? expectContinue, string version)
```
Is this the only test we have for Expect: 100-Continue? Seems like we should at least have a test for the failure case too.

Also, what about other intermediate response codes? I thought we added tests for these when Caesar made all the 1xx changes a while back.

wfurt replied:
no, it is not only one test. But some were not easily generalizable. So I added H2 specific one.

davidsh commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:
```diff
@@ -1344,13 +1345,23 @@ public sealed override async Task<HttpResponseMessage> SendAsync(HttpRequestMess
+                    http2Stream = await SendHeadersAsync(request, cancellationToken).ConfigureAwait(false);
+                    await http2Stream.SendRequestBodyAsync(cancellationToken).ConfigureAwait(false);
+                    // Wait for response headers to be read.
+                    await http2Stream.ReadResponseHeadersAsync().ConfigureAwait(false);
```
The preceding 2 lines pass in the cancellationToken.  But this method doesn't.  Why not?

wfurt replied:
Sending part was fixed by #35118 but ReadResponseHeadersAsync was left out. If we need to plumb cancelation to Read*() functions, we can do that as separate PR. Maybe @stephentoub have some suggestions.

davidsh replied:
A separate PR is fine.  But let's not forget about it for 3.0. We don't want to have gaps in our cancellation logic.

davidsh commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs:
```diff
@@ -194,6 +218,21 @@ public void OnResponseHeadersStart()
+            public void OnResponse100Continue()
+            {
+                // This is called when we receive complete set of headers with status 100 Continue.
+                // We need to rest state and wait for another final response.
```
"rest" -> "reset" ?

davidsh commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs:
```diff
@@ -194,6 +218,21 @@ public void OnResponseHeadersStart()
+                {
+                    if (_state != StreamState.ExpectingData)
+                    {
+                        throw new Http2ProtocolException(Http2ProtocolErrorCode.ProtocolError);
```
@stephentoub Did a recent PR which improved error message and the exceptions.  Is this another place that could benefit from that? I.e. instead of just having `Http2ProtocolErrorCode.ProtocolError` it could include other details in a string or something with the specific reason "expecting more data from stream" (or something like that).

stephentoub replied:
I probably should have done that in general when updating the Htp2ProtocolErrorCode stuff.  If you don't want to do it as part of this PR, I can take a separate pass through as well.

wfurt replied:
I made update to include more info. I can also make separate PR if preferable.

davidsh commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2ProtocolException.cs:
```diff
@@ -16,6 +16,12 @@ public Http2ProtocolException(Http2ProtocolErrorCode protocolError)
         }
 
+        public Http2ProtocolException(string message)
+            : base(message)
```
nit: The `: base(message)` can be on the line above with the rest of the constructor method.

This matches the pattern below. The constructor pattern above is on two lines because putting it on one line would exceed the generally accepted line limit. We break up long lines (generally around 120 characters) to improve readability in the source code.

wfurt replied:
makes sense. updated.

davidsh commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs:
```diff
@@ -221,12 +221,12 @@ public void OnResponseHeadersStart()
                     if (_state != StreamState.ExpectingData)
                     {
-                        throw new Http2ProtocolException(Http2ProtocolErrorCode.ProtocolError);
+                        throw new Http2ProtocolException(SR.Format(SR.net_http_http2_protocol_state, "100Continue", _state));
```
Is the string "100continue" actually received on the wire in the protocol.  Usually we see either a request header "Expect: 100-continue" or the response status line of "100 Continue".

wfurt replied:
It is the second status code. I added space to match it.

wfurt commented:
can you please take another look @stephentoub ? I made few more changes to sync up with recent changes.

wfurt commented:
I did update for #36855 and added mote tests for 100Continue. I also fixed race condition causing deadlock when processing GOAWAY frame. 

gRPC streaming interop passing now:
```
furt@net-dale:/mnt/github/grpc-dotnet/testassets/InteropTestsClient$ /mnt/github/grpc-dotnet/testassets/InteropTestsClient/bin/Debug/netcoreapp3.0/linux-x64/publish/InteropTestsClient  --server_port 8070 --test_case ping_pong
Use TLS: False
Use Test CA: False
Client type: httpclient
Server host: localhost
Server port: 8070
running ping_pong
dbug: Grpc.NetCore.HttpClient.Internal.GrpcCall[1]
      => GrpcMethodType:DuplexStreaming GrpcUri:/grpc.testing.TestService/FullDuplexCall
      Starting gRPC call. Method type: 'DuplexStreaming', URI: '/grpc.testing.TestService/FullDuplexCall'.
dbug: Grpc.NetCore.HttpClient.Internal.GrpcCall[6]
      => GrpcMethodType:DuplexStreaming GrpcUri:/grpc.testing.TestService/FullDuplexCall
      Sending message.
trce: Grpc.NetCore.HttpClient.Internal.GrpcCall[9]
      => GrpcMethodType:DuplexStreaming GrpcUri:/grpc.testing.TestService/FullDuplexCall
      Serialized 'Grpc.Testing.StreamingOutputCallRequest' to 27196 byte message.
trce: Grpc.NetCore.HttpClient.Internal.GrpcCall[7]
      => GrpcMethodType:DuplexStreaming GrpcUri:/grpc.testing.TestService/FullDuplexCall
      Message sent.
trce: Grpc.NetCore.HttpClient.Internal.GrpcCall[2]
      => GrpcMethodType:DuplexStreaming GrpcUri:/grpc.testing.TestService/FullDuplexCall
      Response headers received.
dbug: Grpc.NetCore.HttpClient.Internal.GrpcCall[1]
      => GrpcMethodType:DuplexStreaming GrpcUri:/grpc.testing.TestService/FullDuplexCall
      Reading message.
trce: Grpc.NetCore.HttpClient.Internal.GrpcCall[3]
      => GrpcMethodType:DuplexStreaming GrpcUri:/grpc.testing.TestService/FullDuplexCall
      Deserializing 31423 byte message to 'Grpc.Testing.StreamingOutputCallResponse'.
trce: Grpc.NetCore.HttpClient.Internal.GrpcCall[4]
      => GrpcMethodType:DuplexStreaming GrpcUri:/grpc.testing.TestService/FullDuplexCall
      Received message.
dbug: Grpc.NetCore.HttpClient.Internal.GrpcCall[6]
      => GrpcMethodType:DuplexStreaming GrpcUri:/grpc.testing.TestService/FullDuplexCall
      Sending message.
trce: Grpc.NetCore.HttpClient.Internal.GrpcCall[9]
      => GrpcMethodType:DuplexStreaming GrpcUri:/grpc.testing.TestService/FullDuplexCall
      Serialized 'Grpc.Testing.StreamingOutputCallRequest' to 16 byte message.
trce: Grpc.NetCore.HttpClient.Internal.GrpcCall[7]
      => GrpcMethodType:DuplexStreaming GrpcUri:/grpc.testing.TestService/FullDuplexCall
      Message sent.
dbug: Grpc.NetCore.HttpClient.Internal.GrpcCall[1]
      => GrpcMethodType:DuplexStreaming GrpcUri:/grpc.testing.TestService/FullDuplexCall
      Reading message.
trce: Grpc.NetCore.HttpClient.Internal.GrpcCall[3]
      => GrpcMethodType:DuplexStreaming GrpcUri:/grpc.testing.TestService/FullDuplexCall
      Deserializing 13 byte message to 'Grpc.Testing.StreamingOutputCallResponse'.
trce: Grpc.NetCore.HttpClient.Internal.GrpcCall[4]
      => GrpcMethodType:DuplexStreaming GrpcUri:/grpc.testing.TestService/FullDuplexCall
      Received message.
dbug: Grpc.NetCore.HttpClient.Internal.GrpcCall[6]
      => GrpcMethodType:DuplexStreaming GrpcUri:/grpc.testing.TestService/FullDuplexCall
      Sending message.
trce: Grpc.NetCore.HttpClient.Internal.GrpcCall[9]
      => GrpcMethodType:DuplexStreaming GrpcUri:/grpc.testing.TestService/FullDuplexCall
      Serialized 'Grpc.Testing.StreamingOutputCallRequest' to 1839 byte message.
trce: Grpc.NetCore.HttpClient.Internal.GrpcCall[7]
      => GrpcMethodType:DuplexStreaming GrpcUri:/grpc.testing.TestService/FullDuplexCall
      Message sent.
dbug: Grpc.NetCore.HttpClient.Internal.GrpcCall[1]
      => GrpcMethodType:DuplexStreaming GrpcUri:/grpc.testing.TestService/FullDuplexCall
      Reading message.
trce: Grpc.NetCore.HttpClient.Internal.GrpcCall[3]
      => GrpcMethodType:DuplexStreaming GrpcUri:/grpc.testing.TestService/FullDuplexCall
      Deserializing 2659 byte message to 'Grpc.Testing.StreamingOutputCallResponse'.
trce: Grpc.NetCore.HttpClient.Internal.GrpcCall[4]
      => GrpcMethodType:DuplexStreaming GrpcUri:/grpc.testing.TestService/FullDuplexCall
      Received message.
dbug: Grpc.NetCore.HttpClient.Internal.GrpcCall[6]
      => GrpcMethodType:DuplexStreaming GrpcUri:/grpc.testing.TestService/FullDuplexCall
      Sending message.
trce: Grpc.NetCore.HttpClient.Internal.GrpcCall[9]
      => GrpcMethodType:DuplexStreaming GrpcUri:/grpc.testing.TestService/FullDuplexCall
      Serialized 'Grpc.Testing.StreamingOutputCallRequest' to 45918 byte message.
trce: Grpc.NetCore.HttpClient.Internal.GrpcCall[7]
      => GrpcMethodType:DuplexStreaming GrpcUri:/grpc.testing.TestService/FullDuplexCall
      Message sent.
dbug: Grpc.NetCore.HttpClient.Internal.GrpcCall[1]
      => GrpcMethodType:DuplexStreaming GrpcUri:/grpc.testing.TestService/FullDuplexCall
      Reading message.
trce: Grpc.NetCore.HttpClient.Internal.GrpcCall[3]
      => GrpcMethodType:DuplexStreaming GrpcUri:/grpc.testing.TestService/FullDuplexCall
      Deserializing 58987 byte message to 'Grpc.Testing.StreamingOutputCallResponse'.
trce: Grpc.NetCore.HttpClient.Internal.GrpcCall[4]
      => GrpcMethodType:DuplexStreaming GrpcUri:/grpc.testing.TestService/FullDuplexCall
      Received message.
dbug: Grpc.NetCore.HttpClient.Internal.GrpcCall[1]
      => GrpcMethodType:DuplexStreaming GrpcUri:/grpc.testing.TestService/FullDuplexCall
      Completing client stream.
dbug: Grpc.NetCore.HttpClient.Internal.GrpcCall[1]
      => GrpcMethodType:DuplexStreaming GrpcUri:/grpc.testing.TestService/FullDuplexCall
      Reading message.
trce: Grpc.NetCore.HttpClient.Internal.GrpcCall[2]
      => GrpcMethodType:DuplexStreaming GrpcUri:/grpc.testing.TestService/FullDuplexCall
      No message returned.
dbug: Grpc.NetCore.HttpClient.Internal.GrpcCall[4]
      => GrpcMethodType:DuplexStreaming GrpcUri:/grpc.testing.TestService/FullDuplexCall
      Finished gRPC call.
Passed!
```

wfurt commented:
fixes  #36855

wfurt commented:
test failures seems unrelated. We get stuck on several distros in unrelated test sets. Seems like infrastructure problem.

stephentoub commented on src/System.Net.Http/src/Resources/Strings.resx:
```diff
@@ -446,6 +446,12 @@
+    <value>The HTTP/2 request failed with protocol error '{0}' ({1}).</value>
+  </data>
+  <data name="net_http_http2_protocol_state" xml:space="preserve">
+    <value>HTTP/2 PROTOCOL_ERROR. Received '{0}' in '{1}').</value>
```
Nit: Could you make this into a sentence, e.g.
```C#
<value>The HTTP/2 request encountered a protocol error, receiving '{0}' while in the state '{1}'.</value>
```

stephentoub commented on src/System.Net.Http/src/System/Net/Http/HttpClient.cs:
```diff
@@ -469,6 +469,8 @@ public Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, HttpCompl
                 cts = _pendingRequestsCts;
             }
 
+            request.SendConcurent = completionOption == HttpCompletionOption.ResponseHeadersRead;
```
Nit: Concurent => Concurrent

stephentoub commented on src/System.Net.Http/src/System/Net/Http/HttpRequestMessage.cs:
```diff
@@ -27,6 +27,8 @@ public class HttpRequestMessage : IDisposable
         private bool _disposed;
         private IDictionary<string, object> _properties;
 
+        internal bool SendConcurent;
```
Why do we need this?

geoffkizer replied:
+1. The handler itself shouldn't change behavior based on whether ResponseHeadersRead is specified or not.

wfurt replied:
I was wondering about that. It was not clear if not waiting for sending request can introduce any compatibility issue and it is what we do HTTP/1.1.
It also has best way how to propagate errors to caller IMHO. I can certainly remove this.

stephentoub commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:
```diff
@@ -1198,6 +1199,12 @@ private void AbortStreams(int lastValidStream, Exception abortException)
+
+            if (shouldInvalidate)
+            {
+                // Invalidate outside of lock to avoid race with HttpPool Dispose()
```
It's more about maintaining a lock hierarchy, right? i.e. code holding the pool's lock might call into the connection and try to acquire the connection's lock, which means we mustn't hold the connection's lock while trying to acquire the pool's lock.

wfurt replied:
yes, this is about two treads getting two locks in opposite order. I'll try to get with better explanation.

stephentoub commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:
```diff
@@ -1354,13 +1361,31 @@ public sealed override async Task<HttpResponseMessage> SendAsync(HttpRequestMess
+                    {
+                        await bodyTask.ConfigureAwait(false);
+                        bodyTask = null;
+                    }
```
I'm still not understanding why we need this SendConcurrent.  It seems like an attempt at optimizing the helpers on HttpClient like GetStringAsync?  I'd suggest just getting rid of it, deleting the whole if block above.  You could do an upfront check here if you wanted to see if bodyTask.IsCompleted, and if it is, await it in order to propagate any exceptions before we read the response headers, e.g.
```C#
Task bodyTask = ...;
if (bodyTask.IsCompleted)
{
    await bodyTask.ConfigureAwait(false);
}
```
but otherwise let's just always proceed to read the response.

stephentoub commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:
```diff
@@ -1354,13 +1361,31 @@ public sealed override async Task<HttpResponseMessage> SendAsync(HttpRequestMess
+                    await http2Stream.ReadResponseHeadersAsync().ConfigureAwait(false);
+                    if (bodyTask != null && bodyTask.IsCompleted && bodyTask.IsFaulted)
+                    {
+                        throw bodyTask.Exception;
```
This should be:
```C#
await bodyTask.ConfigureAwait(false);
```

stephentoub replied:
Also, similar to above, we should null out the bodyTask before we await it, so that elsewhere we can avoid processing the bodyTask if it's already bee handled, e.g.
```C#
if (bodyTask != null && bodyTask.IsCompleted)
{
    Task t = bodyTask;
    bodyTask = null;
    await t.ConfigureAwait(false);
}
```

stephentoub commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:
```diff
@@ -1354,13 +1361,31 @@ public sealed override async Task<HttpResponseMessage> SendAsync(HttpRequestMess
+                    }
+                    // Wait for response headers to be read.
+                    await http2Stream.ReadResponseHeadersAsync().ConfigureAwait(false);
+                    if (bodyTask != null && bodyTask.IsCompleted && bodyTask.IsFaulted)
```
You don't need both the IsCompleted and IsFaulted checks here.  In fact, you probably explicitly don't want the IsFaulted check, in case the task ended as Canceled.  Better to just do `if (bodyTask != null && bodyTask.IsCompleted`.

stephentoub commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:
```diff
@@ -1354,13 +1361,31 @@ public sealed override async Task<HttpResponseMessage> SendAsync(HttpRequestMess
+                        bodyTask = null;
+                    }
+                    // Wait for response headers to be read.
+                    await http2Stream.ReadResponseHeadersAsync().ConfigureAwait(false);
```
If this throws an exception, we may end up with `bodyTask` being faulted, in which case we'll want to handle that in the catch block.  That'll probably mean adding code to the catch block that does something like:
```C#
if (bodyTask != null)
{
    bodyTask.ContinueWith((t, state) => Log(state), this, CancellationToken.None, TaskContinuationOptions.OnlyOnFaulted, TaskScheduler.Default);
}
```

stephentoub commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs:
```diff
@@ -92,14 +92,49 @@ public async Task SendRequestBodyAsync(CancellationToken cancellationToken)
+                        }
+
+                        // Don't wait for completion, which could happen asynchronously.
+                        Task ignored = _connection.SendEndStreamAsync(_streamId);
```
Nit: You can replace `Task ignored` with `_`, a C# discard.

stephentoub commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs:
```diff
@@ -92,14 +92,49 @@ public async Task SendRequestBodyAsync(CancellationToken cancellationToken)
+                        // Don't wait for completion, which could happen asynchronously.
+                        Task ignored = _connection.SendEndStreamAsync(_streamId);
+                    }
+                    catch (Exception e) when (e is ObjectDisposedException || e.InnerException is ObjectDisposedException)
```
What's causing us to get ObjectDisposedExceptions?  It'd be better to avoid them from being thrown in the first place.

wfurt replied:
This is from WriteAsync in Http2WriteStream. When we get error code, we abandon and reset stream if we still sending request. I can add better comment or create distinct exception if that make more sense.

stephentoub commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs:
```diff
@@ -92,14 +92,49 @@ public async Task SendRequestBodyAsync(CancellationToken cancellationToken)
+                    await SendRequestBodyAsync(cancellationToken).ConfigureAwait(false);
+                }
+
+                await response;
```
`.ConfigureAwait(false)`

stephentoub commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs:
```diff
@@ -131,15 +166,15 @@ public void OnResponseHeader(ReadOnlySpan<byte> name, ReadOnlySpan<byte> value)
                         if (name.SequenceEqual(s_statusHeaderName))
                         {
-                            if (_response != null)
+                            if (_response != null && (int)_response.StatusCode > 199)
```
What's the extra condition for?

wfurt replied:
I replaced this with new state as @geoffkizer suggested. The original idea was to enforce only one status line but allow 1xx response as they are not final.

stephentoub commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs:
```diff
@@ -131,15 +166,15 @@ public void OnResponseHeader(ReadOnlySpan<byte> name, ReadOnlySpan<byte> value)
                             {
                                 if (NetEventSource.IsEnabled) _connection.Trace("Received duplicate status headers.");
-                                throw new Http2ProtocolException(Http2ProtocolErrorCode.ProtocolError);
+                                throw new Http2ProtocolException(SR.Format(SR.net_http_invalid_response_status_code, "duplicate status"));
```
"duplicate status" => _response.StatusCode

wfurt replied:
It is not the status it self what makes it invalid. 
It is invalid because we got it again. I wanted to capture that information somehow. Any other suggestions?

stephentoub commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs:
```diff
@@ -148,7 +183,7 @@ public void OnResponseHeader(ReadOnlySpan<byte> name, ReadOnlySpan<byte> value)
                                 !IsDigit(status3 = value[2]))
                             {
-                                throw new HttpRequestException(SR.Format(SR.net_http_invalid_response_status_code, Encoding.ASCII.GetString(value)));
+                                throw new Http2ProtocolException(SR.Format(SR.net_http_invalid_response_status_code, Encoding.ASCII.GetString(value)));
```
Why are you changing this exception type?

wfurt replied:
It seems like that is what we throw elsewhere in the code and we warp that in SendAsync(). I think this is some code submitted  in middle of the exception cleanup.

stephentoub commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs:
```diff
@@ -163,15 +198,15 @@ public void OnResponseHeader(ReadOnlySpan<byte> name, ReadOnlySpan<byte> value)
                         {
                             if (NetEventSource.IsEnabled) _connection.Trace("Invalid response pseudo-header '{System.Text.Encoding.ASCII.GetString(name)}'.");
-                            throw new HttpRequestException(SR.net_http_invalid_response);
+                            throw new Http2ProtocolException(SR.net_http_invalid_response);
```
Same question for all of these

stephentoub commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs:
```diff
@@ -345,15 +395,34 @@ public void OnResponseAbort(Exception abortException)
+                }
+                while ((uint)Response.StatusCode < 200);
+
+                if ((int)Response.StatusCode >= 300)
```
Why are we sometimes casting to uint and sometimes casting to int?

wfurt replied:
updated. It will be always positive number but we use int everywhere else.

stephentoub commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs:
```diff
@@ -345,15 +395,34 @@ public void OnResponseAbort(Exception abortException)
             }
 
-            public async Task ReadResponseHeadersAsync()
+            public async Task ReadResponseHeadersAsync(TaskCompletionSource<bool> tcs = null)
```
Please rename `tcs` to something more meaningful, i.e. what does this represent / when should it be completed?

stephentoub commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs:
```diff
@@ -345,15 +395,34 @@ public void OnResponseAbort(Exception abortException)
+                        tcs?.TrySetResult(true);
+                    }
+                }
+                while ((uint)Response.StatusCode < 200);
```
I'm not following this logic... what happens if we receive, e.g., status code == 174... we loop forever?

geoffkizer replied:
I think the logic here is basically correct. Any 1xx response is not a final response, we need to eat it and receive another set of response headers. So if we receive 174, eat it and wait for another set of response headers.

stephentoub commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs:
```diff
@@ -571,6 +640,11 @@ protected override void Dispose(bool disposing)
+                    if (http2Stream._abortException != null)
+                    {
+                        throw http2Stream._abortException;
+                    }
```
This will overwrite any call stack information and Watson information in the exception.  It'd be better to do `ExceptionDispatchInfo.Throw(http2Stream._abortException)`.

stephentoub replied:
Or, actually, shouldn't this be wrapped as the inner exception of an IOException?

wfurt replied:
wrapping happens in SendAsync based on exception type. We may have _abortException because we got RST from server with specific code or because we detected invalid state machine. I think we will raise up IOException if we hit IO error.

stephentoub commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs:
```diff
@@ -582,6 +656,11 @@ public override ValueTask<int> ReadAsync(Memory<byte> destination, CancellationT
 
+                    if (http2Stream._abortException != null)
+                    {
+                        return new ValueTask<int>(Task.FromException<int>(new HttpRequestException(SR.net_http_client_execution_error, http2Stream._abortException)));
```
I think this should be IOException instead of HttpRequestException.  We generally throw IOExceptions out of the read/write operations on these streams, e.g. https://github.com/dotnet/corefx/blob/a28444b71a45171c07eb28a5d82086a4589665fe/src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ContentLengthReadStream.cs#L42

wfurt replied:
updated.

stephentoub commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs:
```diff
@@ -621,10 +700,21 @@ protected override void Dispose(bool disposing)
+                    }
+
+                    // We got rejected while sending request.
+                    if (http2Stream._response != null && (int)http2Stream._response.StatusCode >= 300)
```
Is it possible the response could get nulled out concurrently?  Hopefully not.  If it could, it should be stored into a local and then the local used for both the check and access.

wfurt replied:
no, it should not. We never set it to null once allocated.

geoffkizer replied:
It does seem kind of weird that we are explicitly checking the status code here.

How is this handled in the HTTP/1.1 case?

geoffkizer commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs:
```diff
@@ -345,15 +395,34 @@ public void OnResponseAbort(Exception abortException)
+                    if (Response.StatusCode  == HttpStatusCode.Continue)
+                    {
+                        OnResponse100Continue();
+                        tcs?.TrySetResult(true);
```
Seems like we should set tcs to null here, so we don't try to complete it twice it the unlikely event that we receive more than one 100 continue response.

geoffkizer commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs:
```diff
@@ -216,14 +251,29 @@ public void OnResponseHeadersStart()
+                // We need to reset state and wait for another final response.
+                lock (SyncObject)
+                {
+                    if (_state != StreamState.ExpectingData)
```
It seems weird that we first transition to ExpectingData, then transition *back* to ExpectingHeaders here. Isn't there a timing issue?

Seems like we should check for 1xx while we've got the lock when we are processing headers (in TryEnsureHeaders, I think) and just not transition to ExpectingData if we get a 1xx.

geoffkizer replied:
Actually, TryEnsureHeaders is the wrong place, ignore that. I'll figure out what the right place is.

geoffkizer replied:
The right place to check for 1xx response codes is in `OnResponseHeadersComplete`, I think.

We should check for 1xx status code there under the lock, and if so:
(1) Leave the state as ExpectingHeaders. (Assuming it was already ExpectingHeaders, not ExpectingTrailingHeaders.)
(2) If we have a 100 continue task, complete it and clear it out so it won't be completed again.

We should not need the OnResponse100Continue method at all.

geoffkizer replied:
And there's more we need to do, actually. We need to ignore any headers that come in on a 1xx informational response -- these should not be added to the response object's headers collection. That's how we handle these for HTTP/1.1.

This means we actually need to detect and remember that we saw the 1xx status. This must always be the first header, so it should be relatively straightforward to do.

There's a related issue here (tracked somewhere, I can dig it up if that helps): Status code must always be the first header in the header block. That's good, because it means we can detect 1xx before processing other headers. We don't enforce this today, however.

We should probably split the ExpectingHeaders state into a couple different states:

(1) ExpectingStatus. In this state, we fail if we receive anything other than a status code. If we do, we check for 1xx and transition to one of the states below.
(2) ExpectingHeaders. We received a final response code, so process headers and add to response object.
(3) ExpectingIgnoredHeaders. We received 1xx, so process headers and ignore them.

Then in OnResponseHeadersComplete, ExpectingHeaders will transition to ExpectingData, while ExpectingIgnoredHeaders will transition back to ExpectingStatus.

geoffkizer replied:
There doesn't seem to be a github issue for the status code issue I mentioned above. Regardless, this is something we should address.

geoffkizer replied:
Ah, I see that we are enforcing that status code is always the first header. We do this by checking whether _response is null.

So all we need to do here, I think, is add a status like ExpectingIgnoredHeaders (or whatever name). When we read the status code, if it's 1xx then change the state to ExpectingIgnoredHeaders, and don't create the _response object. When we read a regular header, just ignore it if the state is ExpectingIgnoredHeaders.

wfurt replied:
yes. I'll take a look. that should be easy.

geoffkizer commented on src/System.Net.Http/tests/FunctionalTests/CustomContent.netcore.cs:
```diff
@@ -0,0 +1,54 @@
+            private int _trigger;
+            private byte[] _content;
+            private readonly TaskCompletionSource<bool> _sendingDone;
+            private int _itteration;
```
typo: _itteration => _iteration

geoffkizer commented:
Meta point: There are really two related changes here:

(1) Process 1xx responses correctly (i.e. ignore them and expect another header block). This involves a bunch of changes to the state machine that handles incoming frames, and is related to the issue I mention above about ensuring that status code is always received first. But it doesn't change the overall request logic.
(2) Handle 100-continue logic. If we're sending Expect: 100-continue, then we need to wait to either receive it or have the timer expire before we send the request body. We also need to detect a 100 response and signal the request logic.

These two are fairly separable issues: (2) depends on (1), but (1) could be done first as its own change, and possibly should be, just to simplify and minimize change here.

geoffkizer commented:
I know we added some tests for 1xx informational responses when Caesar implemented the change to handle this properly for HTTP/1.1. Where are these? Shouldn't these be getting enabled for HTTP2?

wfurt commented:
I reworked the state transitions, task handling and addressed feedback. 

It would be great if you can take another look @geoffkizer  and @stephentoub

geoffkizer commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs:
```diff
@@ -120,9 +164,9 @@ public void OnResponseHeader(ReadOnlySpan<byte> name, ReadOnlySpan<byte> value)
                 lock (SyncObject)
                 {
-                    if (_state != StreamState.ExpectingHeaders && _state != StreamState.ExpectingTrailingHeaders)
+                    if (_state == StreamState.ExpectingIgnoredHeaders)
```
It seems like it would be better to do this check below, after we deal with pseudo-headers.

geoffkizer commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs:
```diff
@@ -221,15 +273,19 @@ public void OnResponseHeadersComplete(bool endStream)
                     {
                         _state = StreamState.Complete;
                     }
+                    else if (_state == StreamState.ExpectingIgnoredHeaders)
```
It seems like there's an error case we're not detecting here: 1xx response (i.e. ExpectingIgnoredHeaders) with endStream == true. If this happens, it seems like we will get into a bad state. I think we should detect this and throw a protocol exception.

geoffkizer replied:
I don't think your change fixed the issue. We're checking for endStream above, so we'll never execute the endStream == true case below.

geoffkizer replied:
BTW, it also seems to me like it's an error if _state == ExpectingTrailingHeaders and endStream is false. We're not catching this today, but it seems like we should.

wfurt replied:
updated.

geoffkizer commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs:
```diff
@@ -326,11 +382,13 @@ public void OnResponseAbort(Exception abortException)
-                        Debug.Assert(!_hasWaiter);
-                        _hasWaiter = true;
-                        _waitSource.Reset();
+                        if (!_hasWaiter)
```
Why this change? What makes it so we can get into the case where _hasWaiter is true, when it couldn't be before?

wfurt replied:
What is exact purpose of TryEnsureHeaders() @geoffkizer ? It seems like original logic would check if we have complete headers and if not it would wait once and assume we will never go back and we would proceed with processing body. 

Now with 1xx when we get 100 we get full header set but we will loop and expect another set of header block. I was hitting the assert before as well and that was reason why I would  move back with state instead of adding new state as you suggested.

geoffkizer replied:
If _hasWaiter is true, that means we currently have someone waiting for the headers. That shouldn't be the case here. 

Most likely, I think that we are just not resetting _hasWaiter to false, but I'm not sure how that happens.

Either way, it seems like we shouldn't remove the assert and change the logic here.

wfurt replied:
I'll take another look. I did try to reset _hasWaiter in  OnResponseHeadersComplete() when we finish processing 1xx but that was not sufficient for some reason.

wfurt replied:
The sequence is like this @geoffkizer:
 - we call TryEnsureHeaders() and set awaiter as don't have headers. 
- we wait process headers and call TryEnsureHeaders() again. 
- for 1xx we are in state ExpectingStatus. this is just like when we started. So that creates awaiter. 
- we do 100 processing and loop back. 
- now we call TryEnsureHeaders() again with awaiter set.

We could grab lock and reset that somewhere ReadResponseHeadersAsync() or modify  TryEnsureHeaders() to be smarted and not only look at state but also Response.StatusCode.

Let me know what is your preference.

geoffkizer replied:
I think there's a timing issue here. 

The On* routines (e.g. OnResponseHeader, OnResponseHeadersComplete) get called asynchronously as we process received frames off the connection. They need to be able to update the state of the Http2Stream and response object, and then signal the waiting client task if necessary. We can't control when the client task will actually process the received data, so we need to store all relevant state on the Http2Stream or the response object. The waiter mechanism basically just tells the client task "more data has arrived, so wake up and try to process it".

The problem here is that we used to assume that we only received one header block per response (not counting trailing headers). But now, we can receive multiple header blocks before we receive the body. The logic below in ReadResponseHeadersAsync is assuming that it can process each header block individually. But that's *not* the case, and we shouldn't try to make it the case.

For example:
(1) ReadResponseHeadersAsync calls TryEnsureHeaders and has to wait because headers are available
(2) We receive a header block with status = 100, process it, eventually OnResponseHeadersComplete is called and the waiter will be signalled, but *not necessarily executed immediately*
(3) Now, we receive another header block. Status could be 200 or 103 or 500 or whatever. This will overwrite the existing status on the response object. That's bad.

I think we should change the approach here. TryEnsureHeaders should not signal the waiter for a 1xx response. It should only signal the waiter when it receives the final response headers. This maintains the approach today that TryEnsureHeaders only happens once.

Additionally, we should store the Task for the 100-continue notification on the Http2Stream object, so that when OnResponseHeader receives a 100-continue (or other) status, it can signal the task directly from that code. No need to signal the waiter here, just signal the task directly.

Thoughts?

geoffkizer replied:
BTW, I'm not quite sure how this causes _hasWaiter to be true here, but it seems related.

geoffkizer commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs:
```diff
@@ -621,10 +709,21 @@ protected override void Dispose(bool disposing)
+                    }
+
+                    // We got rejected while sending request.
+                    if (http2Stream._response != null && (int)http2Stream._response.StatusCode >= 300)
```
This is really just a special case of #35652, right?

Seems like what we need here is a general approach to handling exceptions when sending the request body. I.e. not specific to 100-continue.

geoffkizer replied:
I think we need to do a couple things, to address both this specific abort case and the general case of #35652:

(1) Just throw ObjectDisposedException here. We don't need to detect the specific case of 100-continue failure.
(2) Catch exceptions during the send of the request, and send a RST_STREAM at that point. Also, probably set the state to Aborted when this happens.

I think if we do those two things, we'll address all the issues here.

geoffkizer replied:
Also, we need to remove the stream from the stream dictionary when this happens. Not sure if we do this today -- I think we don't.

wfurt replied:
I fixed SendAsync() to dispose AND remove stream just like SendHeadersAsync() does.

wfurt replied:
as far as the 100COntinue: 
What I wanted to do is abandon sending request if we get reject but keep reading response and provide response with code and message to caller rather than throwing IO (or other) exception.

I could move SendRstStreamAsync() to SendAsync(). But I think it would be good to track if we decided to stop sending or the ObjectDisposed is result of protocol or IO error if you agree with behavior described above.

geoffkizer replied:
Yeah, I see your point -- this is not really the same case as other abort cases, because we don't want to fail the request here. Am I understanding you correctly?

Let me make sure I understand the case we're dealing with here...

We already will abort sending the response if we get a >= 300 response before the timer expires, right? We have logic above in ReadResponseHeadersAsync to handle this.

So what case is this addressing? Is it the case where we started sending because the timer expired, and then later received a >= 300 response? For HTTP/1.1, we just finish sending the request body in this case. It seems reasonable to do the same here, for HTTP2. I guess you could argue that it would be better to silently abort the request body in this case, but I'm not sure it matters much. (We can't abort the request body for HTTP/1.1 without killing the connection, so we don't ever do this currently.)

I guess this raises the question of why this stream is disposed at this point? If it wasn't disposed, we'd just keep sending the body in this case and no error would happen.

Note this issue isn't really specific to 100-continue; it's always possible to receive an "early" response, i.e. before we finish sending the request. If we're going to try to abort sending the request in these cases, we should probably do it for non 100-continue cases as well. It's unclear to me if we actually are doing this or not here...

geoffkizer replied:
By the way, I don't see where we are sending RST_STREAM in the case where we get a >= 300 response *before* the timer fires. See my comment in SendRequestBodyWithExpect100ContinueAsync above.

wfurt replied:
> Yeah, I see your point -- this is not really the same case as other abort cases, because we don't want to fail the request here. Am I understanding you correctly?
> 
yes. 


> Let me make sure I understand the case we're dealing with here...
> 
> We already will abort sending the response if we get a >= 300 response before the timer expires, right? We have logic above in ReadResponseHeadersAsync to handle this.
> 
yes. We will not start sending request body if we get negative response before timer expires. 


> So what case is this addressing? Is it the case where we started sending because the timer expired, and then later received a >= 300 response? For HTTP/1.1, we just finish sending the request body in this case. It seems reasonable to do the same here, for HTTP2. I guess you could argue that it would be better to silently abort the request body in this case, but I'm not sure it matters much. (We can't abort the request body for HTTP/1.1 without killing the connection, so we don't ever do this currently.)

yes, we can do that. However the way I read #31312 was that you want us to send abort and stop sending. If we do no stop, wee do not need to send RST, right?

> 
> I guess this raises the question of why this stream is disposed at this point? If it wasn't disposed, we'd just keep sending the body in this case and no error would happen.
> 
> Note this issue isn't really specific to 100-continue; it's always possible to receive an "early" response, i.e. before we finish sending the request. If we're going to try to abort sending the request in these cases, we should probably do it for non 100-continue cases as well. It's unclear to me if we actually are doing this or not here...

One option I was thinking about is simply setting bool flag so we do not need to guess based on exception. Would that be more clean and reasonable?

geoffkizer replied:
> However the way I read #31312 was that you want us to send abort and stop sending. If we do no stop, wee do not need to send RST, right?

Yeah, my comment (2) in the linked issue is about doing exactly this. I don't think it's strictly necessary. It's kind of a corner case: the timer expired so we started sending, but then we received a >= 300 status code. Doesn't seem that common in practice.

So I think we could certainly live without it. But if we're going to add the handling for this, then we may as well handle the same case for non 100-continue as well, i.e. we receive a >= 300 status code while sending the request body.

> One option I was thinking about is simply setting bool flag so we do not need to guess based on exception. Would that be more clean and reasonable?

Yeah, something like this seems reasonable to me. That is, if we receive a final status code while we are in the middle of sending the request body, then:
(1) Set a flag like "_abortRequestBody = true"
(2) Dispose the write stream so it will cause an ObjectDisposedException on the next write
(3) Catch the exception during request body send, and if _abortRequestBody is true, just eat it
(4) We also need to send RST_STREAM when we catch the exception, but we need to do that regardless of whether we are eating the exception or not. This will contribute to #35652, though there may be additional work we need to do to get that working.

geoffkizer commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs:
```diff
@@ -345,15 +409,35 @@ public void OnResponseAbort(Exception abortException)
+                if ((int)Response.StatusCode >= 300)
+                {
+                    // If we tried 100-Continue and got rejected signal that we should not send request body.
+                    shouldSendRequestBody?.TrySetResult(false);
```
Shouldn't we also complete the task (set to true) if the response is < 300? E.g. 200? As it stands, it looks like we'll wait for the timer in this case...

wfurt replied:
original assumption was that if we get final response we are already sending body e.g. timer expired. It is probably cheap so I can do trySetResult always and set value based on result code.

geoffkizer replied:
Yeah, that's probably true, but just in case, it seems reasonable to set the task here anyway.

geoffkizer commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs:
```diff
@@ -92,14 +95,55 @@ public async Task SendRequestBodyAsync(CancellationToken cancellationToken)
+                if (sendRequestContent)
+                {
+                    await SendRequestBodyAsync(cancellationToken).ConfigureAwait(false);
+                }
```
If we're not going to send the content, don't we need to send RST_STREAM here?

wfurt commented:
I re-worked to logic to implement what we discussed in various comments:

- TryEnsureHeaders() logic will block until final headers are received. 
- instead of looping while StatusCode >= 200, we will use new states to skip over 1xx responses
- I added bool _shouldSendRequestBody to replace various places where we tested StatusCode
- this is now set for all responses when we process status code and we use TrySetResult() for 100Continue case
- we should send RST_STREAM now if we do not finish sending request body
- I added more error handling. 

This still may need more testing and I may missed something but it would be nice to get feedback on structural changes. Hopefully this covers all concerns from @geoffkizer

JamesNK commented:
Hi

I'd to be able to test gRPC with this change in the gRPC interop test environment. The interop test environment will test these changes against a dozen different HTTP/2 gRPC server implementations (Java, Go, Node, C++, etc).

How close is this to being merged? It should only take me a day to update the test environment once it is merged, but getting it running there sooner rather than later would allow you to deal with any issues that come up before preview 6 goes out.

stephentoub commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:
```diff
@@ -964,7 +964,7 @@ private void WriteHeaders(HttpRequestMessage request)
         }
 
-        private async ValueTask<Http2Stream> SendHeadersAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+        private async ValueTask<Http2Stream> SendHeadersAsync(HttpRequestMessage request, CancellationToken cancellationToken, bool flush = false)
```
Nit: flush => mustFlush or forceFlush or something like that, just to call out that it might still flush even if this is false.

stephentoub commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:
```diff
@@ -1354,13 +1363,50 @@ public sealed override async Task<HttpResponseMessage> SendAsync(HttpRequestMess
+                        // Sending request body finished before getting headers.
+                        Task t = bodyTask;
+                        bodyTask = null;
+                        await t.ConfigureAwait(false);
```
If this throws, what happens to potential failures in responseHeadersTask?

wfurt replied:
We will call http2Stream.Dispose() bellow and that should make it fail. I'm not sure if you want to propagate the error or if you think we need to so something about it. Right now, I think we will simply throw first failure to caller.

stephentoub replied:
> I'm not sure if you want to propagate the error or if you think we need to so something about it.

If the body task fails, we don't need to also somehow propagate the failure in the responseHeadersTask, but that Task is almost certainly going to also fail with an exception, and if we don't do anything with it, that'll cause TaskScheduler.UnobservedTaskException to be raised.  It'd be better to observe the exception in a continuation and log it... that continuation can be hooked up only in the case in a catch around the await, such that we only pay for it if we're failing anyway.

stephentoub commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:
```diff
@@ -1354,13 +1363,50 @@ public sealed override async Task<HttpResponseMessage> SendAsync(HttpRequestMess
+                        {
+                             _ = bodyTask.ContinueWith((t, state) => {
+                                Http2Connection c = (Http2Connection)state;
+                                if (NetEventSource.IsEnabled) c.Trace($"SendRequestBody Task failed.");
```
We should include t.Exception in the trace

stephentoub commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:
```diff
@@ -1354,13 +1363,50 @@ public sealed override async Task<HttpResponseMessage> SendAsync(HttpRequestMess
 
-                // Wait for response headers to be read.
-                await http2Stream.ReadResponseHeadersAsync().ConfigureAwait(false);
+                    if (bodyTask != null)
```
Instead of `if (bodyTask != null)`, isn't this just `else`?

stephentoub commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:
```diff
@@ -1354,13 +1363,49 @@ public sealed override async Task<HttpResponseMessage> SendAsync(HttpRequestMess
+                    }
+                    else
+                    {
+                        if (bodyTask.IsCompleted)
```
If we get here, the responseHeadersTask completed, right?  We're not awaiting it to propagate any exceptions that may have occurred.

Seems like this logic could be restructured a little, e.g.
```C#
if (bodyTask == await Task.WhenAny(bodyTask, responseHeadersTask).ConfigureAwait(false) ||
    bodyTask.IsCompleted)
{
    // The sending of the request body completed before receiving all of the request headers.
    Task t = bodyTask;
    bodyTask = null;
    try
    {
        await t.ConfigureAwait(false);
    }
    catch
    {
        LogExceptionAsync(this, responseHeadersTask, "Response headers"); // ContinueWith logic factored into a helper
        throw;
    }

    await responseHeadersTask.ConfigureAwait(false);
}
else
{
    // We received the response headers but the request body hasn't yet finished.
    bodyTask.LogExceptionAsync(this, responseHeadersTask, "Request body");
}
```
That said, I'm not clear on the `else` case above where the request headers have completed but the response body hasn't.  If the response body subsequently fails and we're still handling the response body, wouldn't we want to propagate any failure there out via the response body handling?  If that's already completed as well, we're out of luck and logging is the best we can do, but if it hasn't yet completed, seems we'd want to take the opportunity to forward the error out?

wfurt replied:
we don't have problem with ProcessIncomingFramesAsync() because we catch all exceptions, right? If so, can we do something similar and use _abortException?

wfurt replied:
BTW we have few places where we have "ignored" task. Do we have same issue there?

stephentoub replied:
> BTW we have few places where we have "ignored" task. Do we have same issue there?

If exceptions can go unhandled in their implementations, then yes.

stephentoub replied:
> we don't have problem with ProcessIncomingFramesAsync() because we catch all exceptions, right?

Right.

> If so, can we do something similar and use _abortException?

Quite possibly.

wfurt replied:
I updated the code. I did look at what is happening with SendRequestBodyAsync(). If wee get IO exception or if server sends something we do not like (include RST and GOAWAY) we will store that in stream._abortException and we will throw if we still processing response. 
I could explore that more as followup (so as ignored tasks) and we could possibly save all exceptions from SendRequestBodyAsync(). We could also log everything in there to simplify the ContinueWith()

stephentoub replied:
> I could explore that more as followup (so as ignored tasks) and we could possibly save all exceptions from SendRequestBodyAsync()

Yes, let's please follow-up.  There are two concerns:
1. That if an error occurs from sending we don't lose the exception if at all possible.  That means best case propagating it to the response and worst case logging it.
2. That we don't end up with Faulted Tasks that never have their Exceptions observed, which will cause noise for anyone listening to TaskScheduler.UnobservedTaskExceptions.

stephentoub commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs:
```diff
@@ -84,22 +91,68 @@ public Http2Stream(HttpRequestMessage request, Http2Connection connection, int s
+                         _ = _connection.SendRstStreamAsync(_streamId, Http2ProtocolErrorCode.Cancel);
+
+                        // if we decided abandon sending request and we get ObjectDisposed as result of it, just eat exception.
+                        if (_shouldSendRequestBody || (!(e is ObjectDisposedException) && !(e.InnerException is ObjectDisposedException)))
```
How do we know that the ObjectDisposedExceptions we're eating here are being of abandoning sending a request?

wfurt replied:
_shouldSendRequestBody be set to false when we get deny (300+) response code. Originally that was done only for 100Continue but it got changed. Assuming that sending more body will not change server's decision

stephentoub commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs:
```diff
@@ -571,6 +672,11 @@ protected override void Dispose(bool disposing)
                     Http2Stream http2Stream = _http2Stream ?? throw new ObjectDisposedException(nameof(Http2ReadStream));
+                    if (http2Stream._abortException != null)
+                    {
+                        ExceptionDispatchInfo.Throw(new IOException(SR.net_http_client_execution_error, http2Stream._abortException));
```
ExceptionDispatchInfo.Throw is to be used when throwing an exception that was previously thrown and caught, enabling you to avoid overwriting the previous stack trace information.  This is a brand new exception; you can just use throw.

wfurt commented:
/azp run

wfurt commented:
I executed streaming gRPC tests (and few others as well) and it looks OK.

wtgodbe commented:
The test timeout was bogus, the tests completed but never reported back to AzDO. Second or third time I've seen this today (https://github.com/dotnet/corefx/pull/38099#issuecomment-497848710). We should be good to merge this - @wfurt any last concerns?

stephentoub commented:
Let's go ahead and get this in and we can address additional clean-up / fix-up subsequently.

geoffkizer commented:
@stephentoub @wfurt is there an issue for the "additional cleanup" needed here?

wfurt commented:
#36855 is still open @geoffkizer and I plan to use it to close raised issues. I just posed #38226 to do what I discussed with @stephentoub
